### DISCUSSION TITLE
App versioning vs Docker image versioning

### DISCUSSION THREAD

**Discussion by rodrigoechaide** [BretFisher/ama#136] [Q&A] [answered]
Hi Bret, I found this issue (https://github.com/BretFisher/ama/issues/66) that relates to what I am going to ask but that answer not fully solve all my doubts. 
I wanted to ask you what do you think about the best approach to track app version vs docker image version. 

Here goes the questions:

* Is it a good approach to tag the Docker Image with the same version number of the app deployed inside it or it is not? Because I am working on an Project when I am in charge to migrate legacy applications (java apps) to Docker and I am looking for some best practices. I have watched Docker Mastery and Kubernetes Mastery, they are really good and I took a lot of best practices from them but I need some help with this topic. 
* Should I use a separate git repo in order to maintain Dockerfile, configs files and related scripts, in order to deploy app inside Docker image, for each app or should I use the same repo where the source code is maintained? What is the best approach regarding this? 
* And a last doubt do you recommend upload source code artifacts (After code scanning, unit tests and some static analysis) to some repository artifact (such as nexus sonatype) and then from Dockerfile download the artifacts from that registry o just (if source code and Dockerfile are in the same repo) copy all source code to Dockerfile with COPY directive?

Thank you in advance for the answers.

**Comment by BretFisher** [comment]
1. Maybe. I've seen it done many ways. The good thing about keeping versions the same between app and image is it's easy to correlate the two. The bad thing is they don't guarantee you're using the right image. You might change something outside the app, like something in the Dockerfile, or in the CI system itself, and then rebuild. You've now "reused" a tag, which is generally a bad idea in container systems where you'll be automating builds and deployments eventually. I think it's a good place to start though so that you don't make things too complicated in the beginning. Eventually, many teams get to an approach of using git commits for automated image builds, then coming up with a manual tagging approach for app versions on top of that, or maybe a combo like `myapp:3.2-8ab418fe` which is where you manually set the version somewhere in your automation, and the git commit (or CI build ID) is added to the end. Once your CI/CD system is designed to deploy your containers, it'll just use those tags as env vars anyway, but the general rule is:
> For images you plan to push to real servers (test/prod), never reuse a tag
2. Keep dockerfiles in app repos. For compose/kubernetes and other files, a separate DevOps repo might be necessary but keep things as simple as possible, and only add complexity when needed as you grow. The system should be flexible enough so that when you want to split something out, it doesn't take days of rework... hopefully.
3. That sounds like a lot of work. Why not do all that as part of a single Docker build, in the image, or right after the image is built before pushing to a registry? The image *is* the artifact. I don't see a need for two stages of processing, but I haven't seen everything, so maybe there's a reason.

**Comment by rodrigoechaide** [comment, marked as answer]
Thank you very much for the answers Breat! You have been very clear!
